core/src/main/java/org/infinispan/configuration/cache/MemoryConfigurationBuilder.java

```diff
@@ -81,13 +85,24 @@ public MemoryConfigurationBuilder addressCount(int addressCount) {
    public void validate() {
       StorageType type = attributes.attribute(STORAGE_TYPE).get();
       if (type != StorageType.OBJECT && getBuilder().compatibility().isEnabled()) {
-         throw new CacheConfigurationException("Compatibility mode requires OBJECT storage type but was: " + type);
+         throw log.compatibilityModeOnlyCompatibleWithObjectStorage(type);
       }
       long size = attributes.attribute(SIZE).get();
       if (size > 0) {
          EvictionType evictionType = attributes.attribute(EVICTION_TYPE).get();
-         if (evictionType == EvictionType.MEMORY && type != StorageType.BINARY) {
-            throw new CacheConfigurationException("MEMORY based eviction requires binary but was : " + type);
+         if (evictionType == EvictionType.MEMORY) {
+            switch (type) {
+               case OBJECT:
+                  throw log.offHeapMemoryEvictionNotSupportedWithObject();
+               case OFF_HEAP:
+                  int addressCount = attributes.attribute(ADDRESS_COUNT).get();
+                  // Note this is cast to long as we have to multiply by 8 below which could overflow
+                  long actualAddressCount = OffHeapDataContainer.findRequiredMemoryAddress(addressCount);
+                  actualAddressCount *= 8;
+                  if (size < actualAddressCount) {
+                     throw log.offHeapMemoryEvictionSizeNotLargeEnoughForAddresses(size, actualAddressCount, addressCount);
```

Comment:
I don't like the idea that a user could test a configuration on their machine, it would work fine, but then it would fail on a bigger machine. I'd rather make the number of locks smaller than optimal instead of making the address count bigger.

True, the configuration would be suboptimal, but at least it would work the same everywhere.

Reply:
I am not sold on this, personally. In most things you can't have the same exact settings unless the machines themselves have similar specs.